### ISSUE TITLE
Generated task fails when running on Windows with a POSIX-style shell

### ISSUE THREAD

**Issue by Elberet** [microsoft/vscode-eslint#795]
This is a followup to #484 and a re-report of the same issue. I believe that the issue is unrelated to WSL, the reason why it was closed, but I can unfortunately not reply to or reopen the original issue.

The problem:

The auto-supplied eslint task detects the current OS and discovers the installed eslint binary as `node_modules\.bin\eslint.cmd`. It then tries to run that command in the integrated shell. If the integrated shell is any program that does not adhere to Windows path naming conventions - including the WSL bash **but also** any Cygwin based shell or the MinGW/MSYS based `bash` included with Git for Windows - the task will fail.

After a very, very superficial look at the code and the VSCode API I believe that https://github.com/microsoft/vscode-eslint/blob/master/client/src/tasks.ts#L50 is based on an invalid expectation. The command line handed to `ShellExecution` is built to fit the current platform, *but not the selected shell*, which doesn't make sense. If the goal is to find and execute the eslint binary directly, the task should instead use `ProcessExecution`, and only rely on a shell if it cannot find the binary and needs help finding it in the path.

I do have to admit that I have no experience with the inner workings of VSCode and its plugin API, so it may well be that the above doesn't make any sense and the issue remains unfixable in the plugin - but if you do decide to close this again it'd be nice to at least get a better understanding why this is related to WSL. 🤗

---

Environment details:

 - OS: Windows 10 Enterprise 1809
 - Git for Windows: 2.21.0.windows.1
 - VSCode configuration:
   - Version: 1.39.2
   - `"terminal.integrated.shell.windows": "C:\\Program Files\\Git\\bin\\bash.exe"`
   - `"eslint.lintTask.enable": true`
 - Diagnostics via VSCode integrated terminal:
   - `project (master) $ pwd` => `/c/Users/.../project`
   - `project (master) $ bash --version` => `version 4.4.23(1)-release (x86_64-pc-msys)`
   - `project (master) $ ./node_modules/bin/eslint -v` => `v6.6.0`
 - Diagnostics via Windows command prompt:
   - `C:\...\project>bash -c "pwd"` => `/mnt/c/.../project`
   - `C:\...\project>bash --version` => `4.4.20(1)-release (x86_64-pc-linux-gnu)`  
      (Note the differences between the WSL and MSYS bash!)

**Comment by dbaeumer** [comment]
This is a general issue in VS Code and nothing the ESLint extension can fix. See https://github.com/microsoft/vscode/issues/38381

**Comment by Elberet** [comment]
Really? 🤨

Because this task works just fine for me:

```json
    {
      "label": "eslint",
      "type": "process",
      "command": "node_modules\\.bin\\eslint.cmd"
      "args": ["."],
      "problemMatcher": "$eslint-stylish",
      "options": {
        "cwd": "${workspaceFolder}"
      }
    }
```

Isn't that equivalent to using a `ProcessExecution` instead of `ShellExecution` as mentioned in the initial report?

**Comment by Elberet** [comment]
@dbaeumer: looks like you're not reading comments on closed issues. 😁 
Any feedback on the above? Is there a reason for running eslint through a shell that I'm not aware of?

**Comment by dbaeumer** [comment]
I do read comments on closed issue. But I have other things to do as well :-)

In the provided example the OS will very likely spawn a windows command interpreter which executes fine. I talked to the task dev on VS Code and she confirmed that there are still problems executing tasks on bash shells under Windows. T

**Comment by Elberet** [comment]
Hm, apologies... I've accidentally removed a part of the example. Please pretend that the task definition includes `{ "linux": { "command": "./node_modules/.bin/eslint" } }`. Anyway, the significant bit in the sample is the `"type": "process"` setting which I understand causes VS Code to run the command directly instead of invoking a shell with the task command as one of the shell's command line arguments.

This should avoid any problems with executing tasks on bash shells under Windows, because bash is no longer involved at all - or am I missing something else?

**Comment by dbaeumer** [comment]
Yes, but calling the cmd directly doesn't work on all Windows systems. 

One alternative would be to specify the shell to use. Need to check if the API allows this.

As a workaround you can always define your own task.

**Comment by Elberet** [comment]
> calling the cmd directly doesn't work on all Windows systems.

In what way, and why shouldn't the user be expected to provide a sane environment?

(And no, using a POSIX-style integrated shell in Windows is not non-sane, because VS Code explicitly suggests this in its documentation. :wink:)

**Comment by dbaeumer** [comment]
As said this is a problem in VS Code and all I can do is to provide a work around. I can not fix this in ESLint :-)

**Comment by leonbloy** [comment]
For those who have this problem: 
I've solved this by specifying the shell in the task configuration, as explained here 

https://code.visualstudio.com/docs/editor/tasks#_common-questions

In my case I edited .vscode/tasks.json adding the shell thus:


```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "eslint",
			"problemMatcher": [
				"$eslint-stylish"
			],
			"label": "eslint: lint whole folder",
			"options": {
				"shell": {
					"executable": "cmd.exe",
					"args": [
						"/d", "/c"
					]
				}
			}
		}
	]
}
```
